java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -0,0 +1,70 @@
+/**
+ * Provides classes representing various flow steps for taint tracking.
+ */
+
+private import java
+private import semmle.code.java.dataflow.DataFlow
+
+/**
+ * A module importing the frameworks that implement additional flow steps,
+ * ensuring that they are visible to the taint tracking library.
+ */
+module Frameworks {
+  private import semmle.code.java.frameworks.jackson.JacksonSerializability
+  private import semmle.code.java.frameworks.android.Intent
+  private import semmle.code.java.frameworks.android.SQLite
+  private import semmle.code.java.frameworks.Guice
+  private import semmle.code.java.frameworks.Protobuf
+}
+
+/**
+ * A unit class for adding additional taint steps.
+ *
+ * Extend this class to add additional taint steps that should apply to all
+ * taint configurations.
+ */
+class AdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for all configurations.
+   */
+  abstract predicate step(DataFlow::Node node1, DataFlow::Node node2);
+}
+
+/**
+ * A method that preserves taint.
+ *
+ * Extend this class and override at least one of `returnsTaint` or `transfersTaint`
+ * to add additional taint steps through a method that should apply to all taint configurations.
+ */
+abstract class TaintPreservingMethod extends Method {
+  /**
+   * Holds if this method returns tainted data when `arg` tainted.
+   * `arg` is a parameter index, or is -1 to indicate the qualifier.
+   */
+  predicate returnsTaint(int arg) { none() }
+
+  /**
+   * Holds if this method writes tainted data to `sink` when `src` is tainted.
+   * `src` and `sink` are parameter indices, or -1 to indicate the qualifier.
+   */
+  predicate transfersTaint(int src, int sink) { none() }
+}
+
+private class StringTaintPreservingMethod extends TaintPreservingMethod {
+  StringTaintPreservingMethod() {
+    getDeclaringType() instanceof TypeString and
+    hasName(["concat", "copyValueOf", "endsWith", "format", "formatted", "getBytes", "indent",
```

Comment:
Could we add `this.` qualifiers? We're unfortunately not consistent with whether we use this or not, but I think we ought to always include it (to the point where I wish that QL didn't allow one to leave out the qualifier).  This sort of style fix is something that would be a good target for a QL-for-QL query.